### ISSUE TITLE
How to enable determinism?

### ISSUE THREAD

**Issue by yuvaltassa** [SarahWeiii/CoACD#43]
As we attempt to integrate CoACD with MuJoCo, we found that we do not understand how to enable deterministic operation. The [params struct](https://github.com/SarahWeiii/CoACD/blob/fce9d0f63adcb8aae62dae83cd35ebd86645760c/src/config.h#L51) has a `seed` attribute (defaulting to 1234), which seems to imply that rerunning without changing anything will lead to the same decomposition, but this is not the case. Below are images of two runs with identical seed.

Are we doing something wrong? cc @quagla @btaba

<img width="862" alt="Screenshot 2024-05-03 at 12 43 45" src="https://github.com/SarahWeiii/CoACD/assets/6387526/28f68fb8-772a-4dee-96d6-1591970567ab">
<img width="841" alt="Screenshot 2024-05-03 at 12 44 28" src="https://github.com/SarahWeiii/CoACD/assets/6387526/439ec487-cdd1-41a8-9f31-71669ea9a424">

**Comment by SarahWeiii** [comment]
Hi, I think the issue stems from the random seed settings in CDT, but I wasn't able to resolve it before. I'll check it again when I have more time.

**Comment by SarahWeiii** [comment]
Update on this: I audited the remaining nondeterminism sources and fixed two real ones on `main` (fa4fa7f):

1. The merge stage's Hausdorff sampling drew from a thread-local RNG **without reseeding**, so pair costs depended on what the worker thread happened to process before — pure scheduling luck. Sampling is now reseeded per candidate pair.
2. Decomposition results were collected in **thread-completion order**, which fed order-dependent indexing and tie-breaking downstream. Each part now writes to its own slot and results are compacted in a fixed order.

Also for the record: my earlier suspicion about CDT was wrong — the vendored CDT uses a fixed-state SplitMix64 with no system randomness, so it was never the problem.

**What is guaranteed now:** rerunning the same script/binary with the same input mesh and the same `seed` reproduces the same decomposition, and this no longer depends on thread count or scheduling. This holds for the C++ CLI, the Python API, and the Python CLI alike — each is reproducible.

**What is still not guaranteed:** bit-identical results *across different host processes*. We found a reproducible case where the same library, same input, and same seed produce a different (equally valid) decomposition depending on what else is loaded in the calling process — e.g. a Python process that has imported trimesh takes a different search path than the C++ CLI. Both results satisfy the threshold; each environment reproduces its own result consistently. The trigger is pinned down but the root cause is still under investigation.

In short: if you rerun your pipeline, you get your result back. The fixes will be in the next release.
